test/gtest/ucp/test_ucp_request.cc

```diff
@@ -100,3 +102,398 @@ UCS_TEST_P(test_ucp_request, test_request_query)
+
+    void *send_am(std::vector<uint8_t> &sbuf)
+    {
+        ucp_request_param_t req_param = {0};
+        static unsigned am_id         = 1;
+        ucp_am_handler_param_t param;
+        void *sreq;
+
+        param.field_mask = UCP_AM_HANDLER_PARAM_FIELD_ID |
+                           UCP_AM_HANDLER_PARAM_FIELD_CB |
+                           UCP_AM_HANDLER_PARAM_FIELD_ARG;
+        param.id         = am_id;
+        param.cb         = am_data_cb;
+        param.arg        = this;
+
+        ucs_status_t status;
+        status = ucp_worker_set_am_recv_handler(receiver().worker(), &param);
+        ASSERT_UCS_OK(status);
+
+        sreq = ucp_am_send_nbx(sender().ep(), am_id, NULL, 0, sbuf.data(),
+                               sbuf.size(), &req_param);
+        am_id++;
+        return sreq;
+    }
+
+    static void flushed_cb(ucp_request_t *request)
+    {
+        test_proto_reset *self = static_cast<test_proto_reset*>(
+                request->user_data);
+
+        self->m_completed = true;
+        ucp_request_complete_send(request, request->status);
+    }
+
+    static void purge_enqueue_cb(uct_pending_req_t *uct_req, void *arg)
+    {
+        ucp_request_t *req = ucs_container_of(uct_req, ucp_request_t, send.uct);
+        test_proto_reset *self;
+
+        self = static_cast<test_proto_reset*>(arg);
+        self->m_pending.push_back(req);
+    }
+
+    void restart(ucp_ep_h ep)
+    {
+        ucp_request_param_t param;
+        param.op_attr_mask = UCP_OP_ATTR_FIELD_USER_DATA |
+                             UCP_OP_ATTR_FIELD_CALLBACK,
+        param.user_data    = this;
+        param.cb.send      = (ucp_send_nbx_callback_t)ucs_empty_function;
+
+        ucp_ep_purge_lanes(ep, purge_enqueue_cb, this);
+        void *request = ucp_ep_flush_internal(ep, 0, &param, NULL, flushed_cb,
+                                              "ep_restart");
+
+        ASSERT_FALSE(UCS_PTR_IS_ERR(request));
+        if (request != NULL) {
+            wait_for_value(&m_completed, true);
+            ucp_request_release(request);
+        }
+
+        unsigned restart_count = 0;
+
+        UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(ep->worker);
+        for (auto &req : m_pending) {
+            if (req->send.state.dt_iter.offset > 0) {
+                ucp_request_send(req);
+            } else {
+                ucp_proto_request_restart(req);
+                restart_count++;
+            }
+        }
+        UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(ep->worker);
+
+        EXPECT_GT(restart_count, 0);
+    }
+
+    void send_nb(std::vector<uint8_t> &sbuf, mapped_buffer *rbuf,
+                 operation_e op, bool sync, std::vector<void*> &reqs)
+    {
+        ucp_request_param_t param = {0};
+        void *rreq                = NULL;
+        void *sreq                = NULL;
+        ucs::fill_random(sbuf);
+
+        switch (op) {
+        case TAG:
+            param.op_attr_mask = UCP_OP_ATTR_FLAG_NO_IMM_CMPL;
+            if (sync) {
```

Comment:
The args in both cases are same.
Would something like that look better?
`sreq = (sync ? X : Y)(sender().ep(), ..., ...);`